src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -259,417 +257,493 @@ AND NOT EXISTS (
+        if (consumerPredicate != null) {
+            predicates.add(consumerPredicate);
+        }
 
-            ConsumerType ctype = this.consumerTypeCurator.getConsumerType(consumer);
-            if (ctype.isManifest()) {
-                DetachedCriteria hostPoolSubquery = DetachedCriteria.forClass(Pool.class, "PoolI")
-                    .createAlias("PoolI.attributes", "attrib")
-                    .setProjection(Projections.id())
-                    .add(Property.forName("Pool.id").eqProperty("PoolI.id"))
-                    .add(Restrictions.eq("attrib.indices", Pool.Attributes.REQUIRES_HOST));
+        if (ownerId != null) {
+            Join<Pool, Owner> ownerJoin = root.join(Pool_.owner);
+            predicates.add(builder.equal(ownerJoin.get(Owner_.id), ownerId));
+        }
 
-                criteria.add(Subqueries.notExists(hostPoolSubquery));
-            }
-            else if (!consumer.isGuest()) {
-                criteria.add(Restrictions.not(
-                    this.addAttributeFilterSubquery(Pool.Attributes.VIRT_ONLY, Arrays.asList("true"))));
-            }
-            else if (consumer.hasFact(Consumer.Facts.VIRT_UUID)) {
-                String uuidFact = consumer.getFact(Consumer.Facts.VIRT_UUID);
-                Consumer host = null;
+        List<Predicate> poolFilterPredicates = getQualifierPredicates(query, root, builder, product,
+            qualifier);
+        predicates.addAll(poolFilterPredicates);
 
-                if (uuidFact != null) {
-                    host = this.consumerCurator.getHost(uuidFact, ownerId);
-                }
+        query.select(root.get(Pool_.id))
+            .distinct(true)
+            .where(toArray(predicates));
 
-                // Impl note:
-                // This query matches pools with the "requires_host" attribute explicitly set to a
-                // value other than the host we're looking for. We then negate the results of this
-                // subquery, so our final result is: fetch pools which do not have a required host
-                // or have a required host equal to our host.
-
-                // TODO: If we don't have a host, should this be filtering at all? Seems strange to
-                // be filtering pools which have a null/empty required host value. Probably just
-                // wasted cycles.
-                DetachedCriteria hostPoolSubquery = DetachedCriteria.forClass(Pool.class, "PoolI")
-                    .createAlias("PoolI.attributes", "attrib")
-                    .setProjection(Projections.id())
-                    .add(Property.forName("Pool.id").eqProperty("PoolI.id"))
-                    .add(Restrictions.eq("attrib.indices", Pool.Attributes.REQUIRES_HOST))
-                    .add(Restrictions.ne("attrib.elements", host != null ? host.getUuid() : "").ignoreCase());
-
-                criteria.add(Subqueries.notExists(hostPoolSubquery));
-            }
+        List<String> ids = this.getEntityManager()
```

Comment:
I think we have to do the first query to get the max number of Pools to set the `org.candlepin.paging.Page.setMaxRecords()` on the returned Page object. However, maybe we should update `listAvailableEntitlementPools` to return just a list of Pools instead of `Page<List<Pool>>`. I think we have already been converting paged methods to follow that pattern.

Reply:
I think the answer is:
1. Apply the ordering to the query
2. Query all the Pool objects that meet the pool criteria
3. Get the list size as your 'max records' size for the Page object
4. Then filter the List of Pools based on the paging requirements (if they exist)

Reply:
Updated this to be one query instead of 2